### PR TITLE
Make exercise 004 practice

### LINKED ISSUE #2168: Turn exercise 004 into practice
Extract the OCaml code from the markdown file corresponding to this exercise. The file is: `data/exercise/004_NAME.md` (where `004` is the number from the title of this issue and `NAME` is the rest of the file name, for instance `data/exercise/001_tail.md` is the file which corresponds to exercise `001`). Use the code in this file to create a folder `practice/004` (using this same number).

Populate the `004` folder with (almost) the same files as in folders `practice/001` and `practice/002` (they are examples).

Three files are different in each folder:
1. `practice/004/ex.ml`: it contains the function's signature, which needs to be implemented by the learner; 
1. `practice/004/work/impl.ml`: it includes the template meant to be filed by the learner;
1. `practice/004/answer/impl.ml`: consists of the solution from the original markdown file.

The examples included in the original markdown file must be turned into test cases in the `practice/004/ex.ml` file.

Folder structure to create:
```
practice/004/
├── answer
│   ├── impl.ml
│   └── test
│       ├── dune
│       └── run.ml
├── dune
├── dune-project
├── ex.ml
└── work
    ├── impl.ml
    └── test
        ├── dune

### PR DESCRIPTION
**Pull request by Ozyugoo**
Fix PR #2168

### CODE REVIEW

**Comment by Ozyugoo** [comment]
Hi, @cuihtlauac kindly check it to see if it is better now. Thank you

File: practice/004/answer/impl.ml
Comment (cuihtlauac) [inline practice/004/answer/impl.ml:1]: Add newline at end of file

File: practice/004/work/test/run.ml
Comment (cuihtlauac) [inline practice/004/work/test/run.ml:1]: Add newline at end of file

File: practice/004/ex.ml
Comment (cuihtlauac) [inline practice/004/ex.ml:1]: Make sure `dune test --root . answer` is successful. Test on cons fails

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
Congrats @Ozyugoo, you're doing great

**Comment by Ozyugoo** [comment]
> Congrats @Ozyugoo, you're doing great

Thank you @cuihtlauac

File: practice/004/answer/impl.ml
Comment (Ozyugoo) [reply practice/004/answer/impl.ml:1]: Done

File: practice/004/ex.ml
Comment (Ozyugoo) [reply practice/004/ex.ml:1]: I noticed it now, and it is fixed. Thank you

**Comment by Ozyugoo** [comment]
@cuihtlauac I think 'test --root . answer' passed the test now, I noticed I gave a wrong input and expecting a wrong output. Please do check it to confirm.

File: practice/004/ex.ml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (cuihtlauac) [inline practice/004/ex.ml]:
```suggestion
```

File: practice/004/work/test/dune [hunk 2]
```diff
@@ -0,0 +1,5 @@
+(include_subdirs no)
+
+(test
+ (name run)
+ (libraries ounit2 ex))
```
Comment (cuihtlauac) [inline practice/004/work/test/dune]:
```suggestion
(test (name run) (libraries ounit2 ex))
```

File: practice/004/dune [hunk 3]
```diff
@@ -0,0 +1,5 @@
+(include_subdirs qualified)
+
+(library
+ (name ex)
+ (libraries ounit2))
```
Comment (cuihtlauac) [inline practice/004/dune]:
```suggestion
(library (name ex) (libraries ounit2))
```

File: practice/004/answer/impl.ml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+let length list =
+  let rec aux n = function
+    | [] -> n
+    | _ :: t -> aux (n + 1) t
+  in
+  aux 0 list
+
```
Comment (cuihtlauac) [inline practice/004/answer/impl.ml]: You still have an end of file without newline :-)

File: practice/004/answer/test/dune [hunk 5]
```diff
@@ -0,0 +1,5 @@
+(include_subdirs no)
+
+(test
+ (name run)
+ (libraries ounit2 ex))
```
Comment (cuihtlauac) [inline practice/004/answer/test/dune]:
```suggestion
(test (name run) (libraries ounit2 ex))
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
You're mostly ok. Only cosmetic changes

File: practice/004/answer/test/dune (hunk 5 above)
Comment (Ozyugoo) [reply practice/004/answer/test/dune]: Hi @cuihtlauac please help, I effected these changes and tried to run dune test --root . answer 
now am getting this error
dune test --root . answer 
File "dune", line 2, characters 12-15:
2 | (test (name run) (libraries ounit2 ex))
                ^^^
Error: Module "Run" doesn't exist.
File "answer/test/dune", line 2, characters 35-37:
2 | (test (name run) (libraries ounit2 ex))
                                       ^^
Error: Library "ex" not found.
-> required by _build/default/answer/test/run.exe
-> required by alias answer/test/runtest in answer/test/dune:2
did I miss something?

Comment (cuihtlauac) [reply practice/004/answer/test/dune]: Theses commands are working fine on my side with your PR. In which folder are you running them? Have you completed the setup of ocaml.org?

File: practice/004/ex.ml (hunk 1 above)
Comment (cuihtlauac) [reply practice/004/ex.ml]: Don't forget to remove this empty line

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

**Comment by Ozyugoo** [comment]
@cuihtlauac I believe it is completely okay now.

**Review by cuihtlauac — APPROVED** [review]
Great job @Ozyugoo

**Pull request merged**